### ISSUE TITLE
ImageIter much slower than ImageRecordIter

### ISSUE THREAD

**Issue by solin319** [apache/incubator-mxnet#11165]
We use ImageIter to read jpg files,.
batch_size was set to 512. 
MXNET_CPU_WORKER_NTHREADS set to 48
It use about 0.35s per batch.
But it only use about 0.0618s per batch when we use ImageRecordIter to read rec file.

Why ImageIter is too slow?
Does too many augmenters called in python affect this speed?
Does MXNet has an interface can do all the task include read,decode,resize,crop,tanspose,rgb_mean in c++?

**Comment by lanking520** [comment]
Hi @solin319 could you please provide more information about your system information and how you build the package or pip install? Please submit this question on MXNet discussion forum (https://discuss.mxnet.io), where it will get a wider audience and allow other to learn as well.

@nswamy can you add 'Question', 'Performance' tag on this issue?

**Comment by rahul003** [comment]
Performance wise, using a rec file would be much more efficient. Are you trying to avoid using them?

**Comment by solin319** [comment]
Yes,  and I think some users may used to use raw jpeg data like me.
But I can't find a performance method to read.

**Comment by ThomasDelteil** [comment]
@solin319 If you are willing to use Gluon I would suggest combining the `ImageFolderDataset` and `DataLoader` with `num_workers` >> 1 . The data loading is performed asynchronously on several processes and the data is pre-fetched in advance for you.

Have a look at this tutorial: https://mxnet.incubator.apache.org/tutorials/gluon/datasets.html#using-own-data-with-included-datasets

@solin319 Please consider closing if your query has been answered. Thanks!

**Comment by thomelane** [comment]
@solin319 Most augmenters call OpenCV functions written in C++ anyway, so I wouldn't expect much of a difference by moving everything to C++. Just a small overhead from calling in Python.

**Comment by solin319** [comment]
We move some augmenters call in c++, and the speed can achieve 0.21s per batch, which is 0.35 smentioned above.
So we expect community can add an interface can read raw pic with high speed.

**Comment by Ishitori** [comment]
@solin319, have you tried other approaches, like creating your own custom iterator, which can do the reading in the background? I mean, if you use HDD, then using 48 threads to read random images from disk can significantly reduce the speed. Did you try to reduce the number of threads to, for example, 2, to see the difference in loading speed?

Also, how much processing do you do when you read a raw image? Again, maybe you need to have a specialized pipeline: a bunch of threads who are reading files from disk, and another bunch which do processing.
